In [45]:
import json
import numpy as np
import pandas as pd 
import seaborn as sns
from commands import COMMANDS
from matplotlib import pyplot as plt

sns.set_style('whitegrid')

In [46]:
data = pd.read_csv('../data/ad_music_proccessed.csv', index_col=0)
leads = pd.read_csv('../data/leads.csv', index_col=0)

In [47]:
fund_extraction_lead = leads.iloc[0]
fund_extraction_lead[fund_extraction_lead != 0]

cookie                                       32cb902aaf374867c9b0f23a9b0d6dfa
unique_user_ids             ['U486984', 'U734415', 'U557722', 'U734312', '...
num_users                                                                  20
target_artist                                                           [nan]
target_user                                                             [nan]
change_password                                                            20
change_withdrawal_method                                                   20
disable_mfa                                                                20
homepage                                                                   30
personal_details                                                           20
see_balance                                                                56
withdraw_balance                                                           20
Name: 0, dtype: object

In [48]:
ck_data = data[data.cookie.eq(fund_extraction_lead.cookie)]
ck_data = ck_data[ck_data.command.isin(COMMANDS.funds_extraction)]
user_command_count = ck_data.groupby('user_id')['command'].nunique()
qualified_user_ids = user_command_count[user_command_count.eq(4)].index
len(qualified_user_ids)

20

The cookie '32cb902aaf374867c9b0f23a9b0d6dfa' has 20 accounts linked to it,
and has used change_password, change_withdrawal_method, disable_mfa and withdraw_balance  20 times.
from the cell above we learn that each account has used all four commands, so they were all affected.

In [49]:
print(qualified_user_ids)

Index(['U019517', 'U124328', 'U145168', 'U204274', 'U224257', 'U269290',
       'U369562', 'U486984', 'U557722', 'U575442', 'U631730', 'U651065',
       'U734312', 'U734415', 'U748815', 'U750775', 'U770286', 'U880812',
       'U918199', 'U919424'],
      dtype='object', name='user_id')


In [50]:
# Save affected Accounts
out = qualified_user_ids.values.tolist()

try:
    with open("affected_users.json", "r") as f:
        out_file = json.load(f)
except Exception as e:
    out_file = {}

out_file['funds_extraction'] = out
if 'ato' in out_file.keys():
    out_file['ato'] = out_file['ato'] + out
else:
    out_file['ato'] = out
with open("affected_users.json", "w") as f:
    json.dump(out_file, f, indent=4)



In [51]:
total_loss = 0
for user in qualified_user_ids:
    user_data = data[data.user_id.eq(user)]
    user_data = user_data[user_data.cookie.eq('32cb902aaf374867c9b0f23a9b0d6dfa')]
    user_loss = user_data.account_balance.max() - user_data.account_balance.min()
    total_loss += user_loss
    print(f"USER {user} LOST: {user_loss:.2f}$")
print(f"TOTAL LOSS: {total_loss:.2f}$")

USER U019517 LOST: 4869.21$
USER U124328 LOST: 2784.82$
USER U145168 LOST: 9885.61$
USER U204274 LOST: 4695.47$
USER U224257 LOST: 13196.38$
USER U269290 LOST: 13665.74$
USER U369562 LOST: 7876.86$
USER U486984 LOST: 6991.91$
USER U557722 LOST: 10712.65$
USER U575442 LOST: 7271.71$
USER U631730 LOST: 2859.13$
USER U651065 LOST: 14922.02$
USER U734312 LOST: 7198.01$
USER U734415 LOST: 14267.71$
USER U748815 LOST: 3727.69$
USER U750775 LOST: 14279.95$
USER U770286 LOST: 2954.42$
USER U880812 LOST: 11766.07$
USER U918199 LOST: 10615.70$
USER U919424 LOST: 9340.27$
TOTAL LOSS: 173881.33$


Turns out each user has used all four commands with this cookie,
so these 20 users have been affected by the attack.
The total loss is 173,881$.


by applying this query to the entire data we can see other users who have used these four commands.


In [52]:
ck_data = data[data.cookie.eq(fund_extraction_lead.cookie)]
extract = data[data.command.isin(COMMANDS.funds_extraction)]
user_command_count = extract.groupby('user_id')['command'].nunique()
general_user_ids = user_command_count[user_command_count >= 3].index
len(general_user_ids)

23

In [53]:
new_leads = general_user_ids.difference(qualified_user_ids)
new_leads

Index(['U551869', 'U633407', 'U937385'], dtype='object', name='user_id')

U551869 and U633407 sessions' contain no evidence of new cookie, device or ip - i conclude they were unaffected.


U937385 is a new artist account, and has used a single cookie since the creation of his account. I assume he is unaffected (might be related to other attacks.)

I evalutated these accounts by looking at the dataframes below.

In [54]:
data[data.user_id.eq(new_leads[0])]

timestamp  user_id                user_email  \
71727  2024-07-01 17:38:59  U551869  michaeladams@hotmail.com   
71746  2024-07-01 17:50:00  U551869  michaeladams@hotmail.com   
71749  2024-07-01 17:50:38  U551869  michaeladams@hotmail.com   
71756  2024-07-01 17:57:21  U551869  michaeladams@hotmail.com   
71758  2024-07-01 17:59:22  U551869  michaeladams@hotmail.com   
71760  2024-07-01 18:00:29  U551869  michaeladams@hotmail.com   
72546  2024-07-02 12:43:10  U551869  michaeladams@hotmail.com   
72547  2024-07-02 12:45:01  U551869  michaeladams@hotmail.com   
72552  2024-07-02 12:48:29  U551869  michaeladams@hotmail.com   
72553  2024-07-02 12:48:41  U551869  michaeladams@hotmail.com   
72554  2024-07-02 12:48:42  U551869  michaeladams@hotmail.com   
72565  2024-07-02 13:01:11  U551869  michaeladams@hotmail.com   
72566  2024-07-02 13:01:22  U551869  michaeladams@hotmail.com   
72576  2024-07-02 13:11:24  U551869  michaeladams@hotmail.com   
72577  2024-07-02 13:12:37  U551869  michaeladams@hotmail.com   
72578  2024-07-02 13:12:41  U551869  michaeladams@hotmail.com   
72582  2024-07-02 13:15:03  U551869  michaeladams@hotmail.com   
72583  2024-07-02 13:15:31  U551869  michaeladams@hotmail.com   
72587  2024-07-02 13:16:56  U551869  michaeladams@hotmail.com   
72589  2024-07-02 13:19:31  U551869  michaeladams@hotmail.com   
72593  2024-07-02 13:23:39  U551869  michaeladams@hotmail.com   
74807  2024-07-04 12:55:37  U551869  michaeladams@hotmail.com   
74808  2024-07-04 12:56:08  U551869  michaeladams@hotmail.com   
74809  2024-07-04 12:56:32  U551869  michaeladams@hotmail.com   
74810  2024-07-04 12:56:33  U551869  michaeladams@hotmail.com   
74815  2024-07-04 13:00:34  U551869  michaeladams@hotmail.com   
74819  2024-07-04 13:02:46  U551869  michaeladams@hotmail.com   
74827  2024-07-04 13:12:00  U551869  michaeladams@hotmail.com   
74828  2024-07-04 13:13:08  U551869  michaeladams@hotmail.com   
74830  2024-07-04 13:14:06  U551869  michaeladams@hotmail.com   
74831  2024-07-04 13:14:08  U551869  michaeladams@hotmail.com   
74832  2024-07-04 13:14:52  U551869  michaeladams@hotmail.com   

      account_creation_date subscription_status account_tier  \
71727   2021-03-16 00:00:00                paid       artist   
71746   2021-03-16 00:00:00                paid       artist   
71749   2021-03-16 00:00:00                paid       artist   
71756   2021-03-16 00:00:00                paid       artist   
71758   2021-03-16 00:00:00                paid       artist   
71760   2021-03-16 00:00:00                paid       artist   
72546   2021-03-16 00:00:00                paid       artist   
72547   2021-03-16 00:00:00                paid       artist   
72552   2021-03-16 00:00:00                paid       artist   
72553   2021-03-16 00:00:00                paid       artist   
72554   2021-03-16 00:00:00                paid       artist   
72565   2021-03-16 00:00:00                paid       artist   
72566   2021-03-16 00:00:00                paid       artist   
72576   2021-03-16 00:00:00                paid       artist   
72577   2021-03-16 00:00:00                paid       artist   
72578   2021-03-16 00:00:00                paid       artist   
72582   2021-03-16 00:00:00                paid       artist   
72583   2021-03-16 00:00:00                paid       artist   
72587   2021-03-16 00:00:00                paid       artist   
72589   2021-03-16 00:00:00                paid       artist   
72593   2021-03-16 00:00:00                paid       artist   
74807   2021-03-16 00:00:00                paid       artist   
74808   2021-03-16 00:00:00                paid       artist   
74809   2021-03-16 00:00:00                paid       artist   
74810   2021-03-16 00:00:00                paid       artist   
74815   2021-03-16 00:00:00                paid       artist   
74819   2021-03-16 00:00:00                paid       artist   
74827   2021-03-16 00:00:00                paid       artist   
74828   2021-03

In [55]:
data[data.user_id.eq(new_leads[1])]

timestamp  user_id                   user_email  \
40567  2024-07-02 00:49:38  U633407  cummingsmichele@hotmail.com   
40568  2024-07-02 00:54:33  U633407  cummingsmichele@hotmail.com   
40569  2024-07-02 00:57:11  U633407  cummingsmichele@hotmail.com   
40570  2024-07-02 00:57:11  U633407  cummingsmichele@hotmail.com   
40571  2024-07-02 00:57:12  U633407  cummingsmichele@hotmail.com   
40572  2024-07-02 00:59:33  U633407  cummingsmichele@hotmail.com   
40573  2024-07-02 01:00:30  U633407  cummingsmichele@hotmail.com   
40574  2024-07-02 01:01:24  U633407  cummingsmichele@hotmail.com   
40575  2024-07-02 01:04:59  U633407  cummingsmichele@hotmail.com   
40576  2024-07-02 01:05:39  U633407  cummingsmichele@hotmail.com   
40578  2024-07-02 01:07:55  U633407  cummingsmichele@hotmail.com   
40580  2024-07-02 01:12:58  U633407  cummingsmichele@hotmail.com   
42444  2024-07-03 19:55:19  U633407  cummingsmichele@hotmail.com   
42450  2024-07-03 20:03:04  U633407  cummingsmichele@hotmail.com   
42455  2024-07-03 20:10:04  U633407  cummingsmichele@hotmail.com   
42458  2024-07-03 20:11:35  U633407  cummingsmichele@hotmail.com   
42459  2024-07-03 20:12:12  U633407  cummingsmichele@hotmail.com   
42469  2024-07-03 20:23:46  U633407  cummingsmichele@hotmail.com   
42471  2024-07-03 20:26:47  U633407  cummingsmichele@hotmail.com   
42472  2024-07-03 20:30:08  U633407  cummingsmichele@hotmail.com   
42473  2024-07-03 20:34:26  U633407  cummingsmichele@hotmail.com   
42480  2024-07-03 20:37:01  U633407  cummingsmichele@hotmail.com   
42493  2024-07-03 20:44:40  U633407  cummingsmichele@hotmail.com   
42494  2024-07-03 20:45:14  U633407  cummingsmichele@hotmail.com   
42505  2024-07-03 20:52:40  U633407  cummingsmichele@hotmail.com   
42869  2024-07-04 03:51:57  U633407  cummingsmichele@hotmail.com   
42870  2024-07-04 03:52:49  U633407  cummingsmichele@hotmail.com   
42876  2024-07-04 03:54:09  U633407  cummingsmichele@hotmail.com   
42878  2024-07-04 03:55:03  U633407  cummingsmichele@hotmail.com   
42881  2024-07-04 03:56:43  U633407  cummingsmichele@hotmail.com   
42918  2024-07-04 05:06:44  U633407  cummingsmichele@hotmail.com   
42919  2024-07-04 05:07:50  U633407  cummingsmichele@hotmail.com   
42920  2024-07-04 05:08:07  U633407  cummingsmichele@hotmail.com   
42921  2024-07-04 05:11:53  U633407  cummingsmichele@hotmail.com   
42922  2024-07-04 05:13:43  U633407  cummingsmichele@hotmail.com   
42923  2024-07-04 05:13:44  U633407  cummingsmichele@hotmail.com   
42925  2024-07-04 05:15:43  U633407  cummingsmichele@hotmail.com   
42927  2024-07-04 05:17:01  U633407  cummingsmichele@hotmail.com   
42928  2024-07-04 05:17:29  U633407  cummingsmichele@hotmail.com   
42930  2024-07-04 05:22:11  U633407  cummingsmichele@hotmail.com   
42942  2024-07-04 05:31:26  U633407  cummingsmichele@hotmail.com   
42948  2024-07-04 05:33:29  U633407  cummingsmichele@hotmail.com   
42956  2024-07-04 05:38:29  U633407  cummingsmichele@hotmail.com   
42957  2024-07-04 05:38:43  U633407  cummingsmichele@hotmail.com   
44594  2024-07-06 00:04:51  U633407  cummingsmichele@hotmail.com   
44600  2024-07-06 00:11:31  U633407  cummingsmichele@hotmail.com   
44608  2024-07-06 00:31:32  U633407  cummingsmichele@hotmail.com   
44612  2024-07-06 00:35:10  U633407  cummingsmichele@hotmail.com   
44614  2024-07-06 00:37:47  U633407  cummingsmichele@hotmail.com   
44616  2024-07-06 00:39:32  U633407  cummingsmichele@hotmail.com   
45199  2024-07-06 12:24:04  U633407  cummingsmichele@hotmail.com   
45202  2024-07-06 12:27:44  U633407  cummingsmichele@hotmail.com   
45204  2024-07-06 12:28:32  U633407  cummingsmichele@hotmail.com   
45206  2024-07-06 12:28:42  U633407  cummingsmichele@hotmail.com   
45214  2024-07-06 12:32:51  U633407  cummingsmichele@hotmail.com   
45215  2024-07-06 12:32:52  U633407  cummingsmichele@hotmail.com   
45218  2024-07-06 12:35:40  U633407  cummingsmichele@hotmail.com   
45219  2024-07-06 12:37:15  U633407  cummingsmichele@hotmail.com   
45220

In [56]:
data[data.user_id.eq(new_leads[2])][:50]

timestamp  user_id               user_email  \
30969  2024-07-01 08:45:21  U937385  scottholloway@gmail.com   
30978  2024-07-01 08:51:18  U937385  scottholloway@gmail.com   
30981  2024-07-01 08:52:42  U937385  scottholloway@gmail.com   
30984  2024-07-01 08:53:59  U937385  scottholloway@gmail.com   
30989  2024-07-01 09:00:06  U937385  scottholloway@gmail.com   
30990  2024-07-01 09:00:13  U937385  scottholloway@gmail.com   
30994  2024-07-01 09:03:41  U937385  scottholloway@gmail.com   
34443  2024-07-03 19:53:22  U937385  scottholloway@gmail.com   
34444  2024-07-03 19:53:57  U937385  scottholloway@gmail.com   
34445  2024-07-03 19:54:12  U937385  scottholloway@gmail.com   
34448  2024-07-03 19:55:52  U937385  scottholloway@gmail.com   
34449  2024-07-03 19:57:05  U937385  scottholloway@gmail.com   
34450  2024-07-03 19:57:06  U937385  scottholloway@gmail.com   
34452  2024-07-03 19:58:35  U937385  scottholloway@gmail.com   
34455  2024-07-03 19:59:42  U937385  scottholloway@gmail.com   
34456  2024-07-03 19:59:43  U937385  scottholloway@gmail.com   
34458  2024-07-03 20:01:13  U937385  scottholloway@gmail.com   
34460  2024-07-03 20:02:38  U937385  scottholloway@gmail.com   
34461  2024-07-03 20:03:02  U937385  scottholloway@gmail.com   
34463  2024-07-03 20:04:45  U937385  scottholloway@gmail.com   
34464  2024-07-03 20:04:55  U937385  scottholloway@gmail.com   
34469  2024-07-03 20:08:08  U937385  scottholloway@gmail.com   
34475  2024-07-03 20:11:23  U937385  scottholloway@gmail.com   
34892  2024-07-04 07:18:12  U937385  scottholloway@gmail.com   
34895  2024-07-04 07:20:12  U937385  scottholloway@gmail.com   
34897  2024-07-04 07:23:02  U937385  scottholloway@gmail.com   
34898  2024-07-04 07:23:03  U937385  scottholloway@gmail.com   
34900  2024-07-04 07:25:19  U937385  scottholloway@gmail.com   
34901  2024-07-04 07:25:30  U937385  scottholloway@gmail.com   
34902  2024-07-04 07:26:25  U937385  scottholloway@gmail.com   
34903  2024-07-04 07:26:26  U937385  scottholloway@gmail.com   
34908  2024-07-04 07:33:41  U937385  scottholloway@gmail.com   
34912  2024-07-04 07:42:20  U937385  scottholloway@gmail.com   
34913  2024-07-04 07:43:45  U937385  scottholloway@gmail.com   
34914  2024-07-04 07:44:17  U937385  scottholloway@gmail.com   
34916  2024-07-04 07:46:13  U937385  scottholloway@gmail.com   
34917  2024-07-04 07:48:02  U937385  scottholloway@gmail.com   
34929  2024-07-04 07:57:58  U937385  scottholloway@gmail.com   
35071  2024-07-04 09:23:28  U937385  scottholloway@gmail.com   
35107  2024-07-04 09:37:29  U937385  scottholloway@gmail.com   
35117  2024-07-04 09:45:01  U937385  scottholloway@gmail.com   
35120  2024-07-04 09:46:37  U937385  scottholloway@gmail.com   
35126  2024-07-04 09:48:54  U937385  scottholloway@gmail.com   
37006  2024-07-05 22:24:24  U937385  scottholloway@gmail.com   
37010  2024-07-05 22:28:08  U937385  scottholloway@gmail.com   
37013  2024-07-05 22:29:19  U937385  scottholloway@gmail.com   
37014  2024-07-05 22:29:32  U937385  scottholloway@gmail.com   
37015  2024-07-05 22:30:13  U937385  scottholloway@gmail.com   
37016  2024-07-05 22:31:02  U937385  scottholloway@gmail.com   
37017  2024-07-05 22:31:52  U937385  scottholloway@gmail.com   

      account_creation_date subscription_status account_tier  \
30969   2024-07-01 08:45:21                paid       artist   
30978   2024-07-01 08:45:21                paid       artist   
30981   2024-07-01 08:45:21                paid       artist   
30984   2024-07-01 08:45:21                paid       artist   
30989   2024-07-01 08:45:21                paid       artist   
30990   2024-07-01 08:45:21                paid       artist   
30994   2024-07-01 08:45:21                paid       artist   
34443   2024-07-01 08:45:21                paid       artist   
34444   2024-07-01 08:45:21                paid       artist   
34445   2024-07-01 08:45:21                paid       artist   
34448   2024-07-01 08:45:21                paid 